config/asrockrack.go

```diff
@@ -0,0 +1,112 @@
+package config
+
+import (
+	"encoding/xml"
+	"strings"
+)
+
+type asrockrackVendorConfig struct {
+	ConfigFormat string
+	ConfigData   *asrockrackConfig
+}
+
+type asrockrackConfig struct {
+	BiosCfg *asrockrackBiosCfg `xml:"BiosCfg"`
+}
+
+type asrockrackBiosCfg struct {
+	XMLName xml.Name                 `xml:"BiosCfg"`
+	Menus   []*asrockrackBiosCfgMenu `xml:"Menu"`
+}
+
+type asrockrackBiosCfgMenu struct {
+	XMLName  xml.Name                    `xml:"Menu"`
+	Name     string                      `xml:"name,attr"`
+	Settings []*asrockrackBiosCfgSetting `xml:"Setting"`
+	Menus    []*asrockrackBiosCfgMenu    `xml:"Menu"`
+}
+
+type asrockrackBiosCfgSetting struct {
+	XMLName        xml.Name `xml:"Setting"`
+	Name           string   `xml:"Name,attr"`
+	Order          string   `xml:"order,attr"`
+	SelectedOption string   `xml:"selectedOption,attr"`
+	Type           string   `xml:"type,attr"`
+}
+
+func NewAsrockrackVendorConfigManager(configFormat string) (VendorConfigManager, error) {
+	asrr := &asrockrackVendorConfig{}
+
+	switch strings.ToLower(configFormat) {
+	case "json":
+		asrr.ConfigFormat = strings.ToLower(configFormat)
+	default:
+		return nil, UnknownConfigFormatError(strings.ToLower(configFormat))
+	}
+
+	asrr.ConfigData = &asrockrackConfig{
+		BiosCfg: &asrockrackBiosCfg{},
+	}
+
+	return asrr, nil
+}
+
+// FindMenu locates an existing asrockrackBiosCfgMenu if one exists in the ConfigData, if not
+// it creates one and returns a pointer to that.
+func (cm *asrockrackVendorConfig) FindMenu(menuName string) (m *asrockrackBiosCfgMenu) {
+	for _, m = range cm.ConfigData.BiosCfg.Menus {
+		if m.Name == menuName {
+			return
+		}
+	}
+
+	m.Name = menuName
+
+	cm.ConfigData.BiosCfg.Menus = append(cm.ConfigData.BiosCfg.Menus, m)
+
+	return
+}
+
+// FindMenuSetting locates an existing asrockrackBiosCfgSetting if one exists in the
+// ConfigData, if not it creates one and returns a pointer to that.
+func (cm *asrockrackVendorConfig) FindMenuSetting(m *asrockrackBiosCfgMenu, name string) (s *asrockrackBiosCfgSetting) {
+	for _, s = range m.Settings {
+		if s.Name == name {
+			return
+		}
+	}
+
+	s.Name = name
+
+	m.Settings = append(m.Settings, s)
```

Comment:
In response to all of these related comments:

If m.Settings is nil I imagine the append would fail, but if it was 'empty' it would work and m.Settings would no longer be empty afterwards.   Is that correct?

Reply:
Thats correct, and I also now realise the values are actually being initialised on line 47.
Ignore my comments on this subject.